Lib/test/test_sax.py

```diff
@@ -1339,6 +1340,169 @@ def test_nsattrs_wattr(self):
+        self.test_data.write('<body>Teach me patience!</body>\n')
+        self.test_data.write('<footer>&{};&{};&{};</footer>\n'.
+                             format(self.specified_entity_names[1],
+                                    self.specified_entity_names[0],
+                                    self.specified_entity_names[2]))
+        self.test_data.write('<!-- {} -->\n'.format(self.specified_comment[1]))
+        self.test_data.write('</{}>\n'.format(self.specified_doctype))
+        self.test_data.seek(0)
+
+        # Data received from handlers - to be validated
+        self.version = None
+        self.encoding = None
+        self.standalone = None
+        self.doctype = None
+        self.publicID = None
+        self.systemID = None
+        self.end_of_dtd = False
+        self.comments = []
+
+    def test_handlers(self):
+        class TestLexicalHandler(LexicalHandler):
+            def __init__(self, test_harness, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+                self.test_harness = test_harness
+
+            def xmlDecl(self, version, encoding, standalone):
+                self.test_harness.version = version
+                self.test_harness.encoding = encoding
+                self.test_harness.standalone = standalone
+
+            def startDTD(self, doctype, publicID, systemID):
+                self.test_harness.doctype = doctype
+                self.test_harness.publicID = publicID
+                self.test_harness.systemID = systemID
+
+            def endDTD(self):
+                self.test_harness.end_of_dtd = True
+
+            def comment(self, text):
+                self.test_harness.comments.append(text)
+
+        self.parser = create_parser()
+        self.parser.setContentHandler(ContentHandler())
+        self.parser.setProperty(
+            'http://xml.org/sax/properties/lexical-handler',
+            TestLexicalHandler(self))
+        source = InputSource()
+        source.setCharacterStream(self.test_data)
+        self.parser.parse(source)
+        self.assertEqual(self.version, self.specified_version)
+        self.assertEqual(self.encoding, self.specified_encoding)
+        self.assertIsNotNone(self.standalone)
+        self.assertEqual(self.doctype, self.specified_doctype)
+        self.assertIsNone(self.publicID)
+        self.assertIsNone(self.systemID)
+        self.assertTrue(self.end_of_dtd)
+        for i in range(1):
+            self.assertEqual(len(self.comments),
+                             len(self.specified_comment))
+            self.assertEqual(' {} '.format(self.specified_comment[i]),
+                             self.comments[i])
+
+
+class CDATAHandlerTest(unittest.TestCase):
+    """This is implemented as a separate class since CDATA sections in XML
+    cannot appear within elements defined within a DTD. Hence this test does
+    not use a DTD."""
+    def __init__(self, *args, **kwargs):
```

Comment:
Doing something in `TestCase.__init__()`, especially setting up test data, is a very unusual thing to do. Please use the normal set-up methods instead.